### ISSUE TITLE
GSoC ideas - Zotonic

### ISSUE THREAD

**Issue by mworrell** [erlef/gsoc#2]
**Title**

Zotonic - Elixir integration

**Description**

Zotonic is the Erlang Content Management System and Framework.

It provides a complete environment to build web sites, with integrated administrative
environment, module system, virtual hosting, users, media, content management,
templates, emailing, LetsEncrypt, etc.

We would like to make modules and sites with Elixir.

For this:

* The build system (rebar3) must know how to handle Elixir applications;
* The runtime module and site managers must be able to introspect Elixir modules;
* The template compiler must be able to call into Elixir modules for filters and such;
* The MQTT routines must understand models implemented in Elixir;
* The notification system must understand observers written in Elixir;
* The `zotonic_filehandler` must understand how to compile Elixir files and/or projects.

And probably some other changes here and there to make it a smooth experience for Elixir developers.

**Recommended Skills**

Experience with Elixir and Erlang.  You must be willing to investigate how Elixir functions
are called from Erlang and dive into the runtime behavior of the Beam.

**Mentor(s)**

Marc Worrell  @mworrell
Maas-Maarten Zeeman  @mmzeeman
With help from Arjan Scherpenisse  @arjan

**GitHub project url**

https://github.com/zotonic/zotonic

**Comment by mworrell** [comment]
Hi @benoitc, is this mergeable/copyable as a project?

**Comment by benoitc** [comment]
i was offline due to a network issue. I will add you , though i didn't
check the calendar yet but student have already started to write their
proposals

**Comment by benoitc** [comment]
Hrm in fact it may be too late unless you have already a student in mind :/

Here is the schedule:

[image: Screenshot 2020-03-23 at 15.51.32.png]

Let me know if you still want to proceed,

Best,

Benoît

**Comment by mworrell** [comment]
if you could merge it - then we have it for next year already :-)

**Comment by mworrell** [comment]
We are a bit busy - so delayed with writing this :-(

**Comment by benoitc** [comment]
arghhh I missed your answer... Was sick. So not sure what needed to be
done. I understood it was not needed... But unsure now :/ I can still put
it on the wiki but does it worth it?

Benoît

**Comment by mworrell** [comment]
If it is no too much work for you. When it is there I will ask around if people have some students that are interested.

If all fails then we have a good project on the wiki for next year.

Hope you are feeling better now!

**Comment by benoitc** [comment]
if you want to participate this year, please proceed with an updated ticket :)

**Comment by mworrell** [comment]
Thanks!  I think this project is still very relevant.

/cc @mmzeeman @michielklonhammer

**Comment by mworrell** [comment]
@arjan, could you check this GSoC idea?

**Comment by arjan** [comment]
Yep I think this is indeed still relevant and actually also pretty doable.

**Comment by mworrell** [comment]
@benoitc this is still good to go! Would be great to have someone working in this.

**Comment by mworrell** [comment]
@benoitc Did you see the previous message?  We are happy with this project description, and it is good to go.

**Comment by pablocostass** [comment]
Hey @mworrell, I totally forgot to ping you yesterday but I went ahead and added the project idea from this issue to the list of proposals! :)

**Comment by mworrell** [comment]
@pablocostass Thanks!  Hope we will find someone for this project.

**Comment by pablocostass** [comment]
This seems to have become a yearly tradition, but @mworrell, if we get accepted for this year's GSoC, do you want to still have the [Zotonic proposal](https://github.com/erlef/gsoc/wiki/Project:-Zotonic) on the list of ideas?

**Comment by arjan** [comment]
:+1: 

I think with Zotonic 1.x which is more OTP-oriented, this integraton will become a lot simpler. At least, having separate zotonic modules (which are now OTP applications) as Elixir projects seems perfectly possible using [rebar_mix](https://hex.pm/packages/rebar_mix).

**Comment by pablocostass** [comment]
So it stays, nice!

**Comment by mworrell** [comment]
Yes please! It is actually still on our roadmap/wishlist. So would be great to find someone who can work on this.

**Comment by pablocostass** [comment]
@mworrell @arjan so it seems that for this year's edition we have to add the following information to the project, would you be so kind to get back to me with it?
- expected outcome
- expected size of project (175 or 350 hour) 
- an easy, medium or hard difficulty rating of each project

**Comment by mworrell** [comment]
## Expected outcome

An extension to Zotonic that makes it possible to use Elixir OTP applications as modules and sites.

## Expected size

350

@arjan, what do you think?

I estimated:

 - 1 week (40 hours) to get to know the system
 - 1 week (40 hours) figuring our Elixir / Erlang interop
 - 1 week (40 hours) getting elixir projects compiled, first introspection. draft of a plan
 - 1 week (40 hours) complete introspection, changing notifiers etc to hook into Erlang
 - 1 week (40 hours) miscellaneous / production ready / bug fixes
 - 1 week (40 hours) extra test sets
 - 1 week (40 hours) demo project, documentation

Total: 280 hours.

I estimate more hours than I would for one of us, as the student had to learn on the job.
I think 175 is a bit too short to get this done, especially with the learning involved.

## Rating

Medium

**Comment by harshalmittal4** [comment]
Hi @mworrell , 
I am currently learning Elixir and am excited about its future. Could you please tell if erlef will be participating in GSoC 2023. Thanks

**Comment by mworrell** [comment]
Hi @harshalmittal4, nice to hear that you are interested in Elixir and the beam (cq Erlang). Every year we try to join the GSoC, but there are more organizations that want to join than spots made available by Google. Last year we regretfully didn't get a spot. Let's hope that this year will be different.

**Comment by harshalmittal4** [comment]
Thanks @mworrell for replying. 
If erlef gets selected this year could you please also tell what will some of the ideas be for this year.

**Comment by mworrell** [comment]
@harshalmittal4 Definitely the project above.  We are very interested in getting Elixir play nice with Zotonic.

Besides that we have many other possible projects, examples are Semantic Web data federation, Matrix support, fail-over handling, real-time communication and much more.

**Comment by harshalmittal4** [comment]
I will get a high level understanding of the codebase and hope the project gets selected for GSoC!
